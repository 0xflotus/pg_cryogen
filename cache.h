#ifndef __CACHE_H__
#define __CACHE_H__

#include "postgres.h"
#include "storage/block.h"
#include "utils/relcache.h"

#include "scan_iterator.h"


#define InvalidCacheEntry -1

typedef enum
{
    CRYO_ERR_SUCCESS = 0,
    CRYO_ERR_DECOMPRESSION_FAILED,
    CRYO_ERR_WRONG_STARTING_BLOCK,
    CRYO_ERR_CACHE_IS_FULL
} CryoError;

/* Position in cache */
typedef int CacheEntry;

void cryo_init_cache(void);
CryoError cryo_read_data(Relation rel, SeqScanIterator *iter, BlockNumber block,
                         CacheEntry *result);
CacheEntry cryo_cache_allocate(Relation rel, BlockNumber blockno);
void cryo_cache_release(CacheEntry entry);

uint32 cryo_cache_get_pg_nblocks(CacheEntry entry);
char *cryo_cache_get_data(CacheEntry entry);
TransactionId cryo_cache_get_xid(CacheEntry entry);
char *cryo_cache_err(CryoError err);


#endif /* __CACHE_H__ */
