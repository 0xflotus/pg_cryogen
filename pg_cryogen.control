comment = 'compressed heap tuple storage'
default_version = '0.2'
module_pathname = '$libdir/pg_cryogen'
relocatable = true
