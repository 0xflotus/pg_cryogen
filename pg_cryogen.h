#ifndef __PG_CRYOGEN_H__
#define __PG_CRYOGEN_H__

#include "access/tableam.h"
#include "utils/snapshot.h"


extern const TableAmRoutine CryoAmRoutine;

extern bool xid_is_visible(Snapshot snapshot, TransactionId xid);

#endif
