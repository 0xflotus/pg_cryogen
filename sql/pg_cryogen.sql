CREATE EXTENSION pg_cryogen;

CREATE TEMPORARY TABLE tmp_data
AS SELECT gs as id, md5(gs::TEXT) as msg FROM generate_series(1, 500) as gs;

COPY tmp_data TO '/tmp/pg_cryogen_data.csv' WITH csv;

CREATE TABLE data (id INT4 NOT NULL, msg TEXT) USING pg_cryogen;
COPY data FROM '/tmp/pg_cryogen_data.csv' WITH csv;

SELECT * FROM data LIMIT 10;
SELECT avg(id) FROM data;
SELECT count(*) FROM data;

CREATE INDEX data_id_idx ON data USING btree (id);
SET enable_seqscan = f;
ANALYZE data;
EXPLAIN (COSTS OFF) SELECT * FROM data WHERE id = 500;
SELECT * FROM data WHERE id = 500;
DROP INDEX data_id_idx;
CREATE INDEX data_id_idx ON data USING brin (id);
EXPLAIN (COSTS OFF) SELECT * FROM data WHERE id = 500;
SELECT * FROM data WHERE id = 500;
SET enable_seqscan = t;

SET pg_cryogen.compression_method = 'lz4';
COPY data FROM '/tmp/pg_cryogen_data.csv' WITH csv;
SELECT count(*) FROM data;

VACUUM data;

-- inserts
TRUNCATE data;
INSERT INTO data
SELECT gs as id, md5(gs::TEXT) as msg FROM generate_series(1, 500) as gs;

-- check FKs
DROP INDEX data_id_idx;
CREATE UNIQUE INDEX data_id_idx ON data USING btree (id);
CREATE TABLE ref_table (id SERIAL, data_id INT4 REFERENCES data (id));
INSERT INTO ref_table VALUES (1, 50), (2, 100), (3, 250); -- cryo_tuple_lock is called here
EXPLAIN (COSTS OFF) SELECT * FROM data JOIN ref_table ON data_id = data.id;
SELECT * FROM data JOIN ref_table ON data_id = data.id LIMIT 3;
DROP TABLE ref_table;
DROP INDEX data_id_idx;

-- check rescan
CREATE TABLE other_table AS
SELECT gs as id, '2000-01-01'::DATE + gs as dt FROM generate_series(1, 500) as gs;
SET enable_hashjoin = f;
SET enable_mergejoin = f;
SET enable_material = f;
EXPLAIN (COSTS OFF) SELECT * FROM data JOIN other_table USING (id);
SELECT * FROM data JOIN other_table USING (id) LIMIT 3;
