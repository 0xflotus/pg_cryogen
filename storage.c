#include "storage.h"


void
cryo_init_page(CryoDataHeader *hdr)
{
    memset(hdr, 0, CRYO_BLCKSZ);
    hdr->lower = CryoDataHeaderSize;
    hdr->upper = CRYO_BLCKSZ;
}

bool
cryo_storage_insert(CryoDataHeader *d, HeapTuple tuple)
{
    CryoItemId  itemId;

    /* check there is enough space */
    if ((tuple->t_len + sizeof(ItemId)) > (d->upper - d->lower))
    {
        /* not enough space */
        return false;
    }

    /* insert tuple */
    d->upper -= MAXALIGN(tuple->t_len);
    memcpy((char *) d + d->upper, tuple->t_data, tuple->t_len);

    /* insert item id pointing to the tuple */
    itemId.off = d->upper;
    itemId.len = tuple->t_len;
    memcpy((char *) d + d->lower, &itemId, sizeof(ItemId));
    d->lower += sizeof(ItemId);

    return true;
}
