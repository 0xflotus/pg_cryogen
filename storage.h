#ifndef __STORAGE_H__
#define __STORAGE_H__

#include "postgres.h"
#include "access/htup.h"
#include "storage/bufpage.h"

#include "compression.h"


#define STORAGE_VERSION     1

/*
 * XXX probably it makes sense to make page size configurable for each table
 * but that would require more complex solution for cache
 */
#define CRYO_BLCKSZ (1 << 20)   /* 1Mb */

typedef struct
{
    PageHeaderData  base;           /* to keep PageIsVerified quiet */
    uint16          version;        /* storage version */
    uint32          target_block;   /* the last block we inserted to;
                                     * zero if we haven't yet */
    uint64          ntuples;        /* total number of tuples in relation */
} CryoMetaPage;

/*
 * Compressed cryo page maps to several postgres blocks. Each block has
 * a simple header containing block number relative to the starting one.
 */
typedef struct
{
    PageHeaderData  base;           /* we don't use it, but it is required by
                                     * GenericXLogFinish() */
    //uint16          curpage;
    BlockNumber     first;
    BlockNumber     next;
} CryoPageHeader;

/*
 * First cryo page also contains additional metadata on compressed cryo block.
 */
typedef struct
{
    CryoPageHeader  cryo_base;
    TransactionId   created_xid;    /* transaction performed insertion */
    CompressionMethod compression_method;
    uint32          compressed_size;
    uint16          npages;         /* number of pages for this cryo block */
} CryoFirstPageHeader;

#define CryoPageHeaderSize(page, block) \
    ((page)->first == block ? sizeof(CryoFirstPageHeader) : sizeof(CryoPageHeader))

/* */
typedef struct
{
    uint32  off;
    uint32  len;
} CryoItemId;

typedef struct
{
    uint32  lower;
    uint32  upper;
    char    data[];
} CryoDataHeader;

#define CryoDataHeaderSize offsetof(CryoDataHeader, data)


void cryo_init_page(CryoDataHeader *hdr);
int cryo_storage_insert(CryoDataHeader *d, HeapTuple tuple);
HeapTuple cryo_storage_fetch(CryoDataHeader *d, int pos, HeapTuple tuple);


#endif /* __STORAGE_H__ */
