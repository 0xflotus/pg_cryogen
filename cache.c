#include <sys/time.h>

#include "cache.h"
#include "compression.h"
#include "storage.h"

#include "access/visibilitymap.h"
#include "storage/buf.h"
#include "storage/bufmgr.h"
#include "utils/catcache.h"
#include "utils/hsearch.h"
#include "utils/rel.h"


/* TODO: make it configurable via GUC variable */
#define CACHE_SIZE 16

/* TODO: concatenate headers and put this definition in it */
#define MIN(a, b) ((a) < (b) ? (a) : (b))

typedef struct
{
    Oid         relid;
    BlockNumber blockno;
} PageId;

typedef struct
{
    PageId      key;
    CacheEntry  entry;
} PageIdCacheEntry;

/* Timestamp */
typedef unsigned long long TS;

typedef struct
{
    PageId      key;
    TS          ts;         /* timestamp for LRU */
    uint32      nblocks;    /* number of postgres blocks */
    TransactionId xid;      /* transaction created block */
    char        data[CRYO_BLCKSZ];
} CacheEntryHeader;

/* Cache itself */
CacheEntryHeader cache[CACHE_SIZE];

/* Hashtable for quick cache lookups by (relid, blockno) key */
HTAB *pagemap;

void
cryo_init_cache(void)
{
    HASHCTL     hash_ctl;

    /* Initialize hash tables used to track TIDs */
    memset(&hash_ctl, 0, sizeof(hash_ctl));
    hash_ctl.keysize = sizeof(PageId);
    hash_ctl.entrysize = sizeof(PageIdCacheEntry);
    hash_ctl.hcxt = CacheMemoryContext;

    pagemap =
        hash_create("pg_cryogen cache",
                    CACHE_SIZE,
                    &hash_ctl,
                    HASH_ELEM | HASH_BLOBS | HASH_CONTEXT);
}

/*
 * Return current UNIX time in microseconds
 */
static TS
get_current_timestamp_ms(void)
{
    struct timeval tv;

    gettimeofday(&tv, NULL);

    return (TS)(tv.tv_sec) * 1000 +
           (TS)(tv.tv_usec) / 1000;
}

/*
 * Read and reassemble compressed data from the buffer manager starting
 * with `block` and decompress it.
 *
 * Possible return values:
 *      CRYO_ERR_SUCCESS: success;
 *      CRYO_ERR_WRONG_STARTING_BLOCK: specified blockno is not a starting
 *          block of cryo page;
 *      CRYO_ERR_DECOMPRESSION_FAILED: decompression failure.
 *
 * XXX Move visibility check here to avoid unnecessary i/o.
 */
static CryoError
cryo_read_decompress(Relation rel, BlockNumber block, CacheEntryHeader *entry)
{
    Buffer      buf;
    Buffer      vmbuf = InvalidBuffer;
    CryoPageHeader *page;
    char       *compressed, *p;
    Size        compressed_size, size;
    uint8       vmflags;

    buf = ReadBuffer(rel, block);
    page = (CryoPageHeader *) BufferGetPage(buf);

    /*
     * In case of brin index bitmap scan can try to read data not from the
     * first page.
     */
    if (page->curpage != 0)
    {
        ReleaseBuffer(buf);
        return CRYO_ERR_WRONG_STARTING_BLOCK;
    }

    size = compressed_size = ((CryoFirstPageHeader *) page)->compressed_size;
    p = compressed = palloc(compressed_size);
    entry->nblocks = 1;

    /*
     * Check whether page is frozen.
     *
     * XXX We don't actually write FrozenTransactionId into page header as this
     * would require entire page copy to WAL (due to Generic WAL), and VACUUM
     * would be very write heavy. Instead we just write frozen bit into
     * visibility map.
     */
    vmflags = visibilitymap_get_status(rel, block, &vmbuf);
    entry->xid = (vmflags & VISIBILITYMAP_ALL_FROZEN) ?
        FrozenTransactionId : ((CryoFirstPageHeader *) page)->created_xid;
    if (BufferIsValid(vmbuf))
        ReleaseBuffer(vmbuf);

    while (true)
    {
        Size    content_size = BLCKSZ - CryoPageHeaderSize(page);
        Size    l = MIN(content_size, size);
        char   *page_content = ((char *) page) + CryoPageHeaderSize(page);

        memcpy(p, page_content, l);
        p += l;
        size -= l;
        ReleaseBuffer(buf);

        if (size == 0)
            break;

        /* read the next block */
        buf = ReadBuffer(rel, ++block);
        page = (CryoPageHeader *) BufferGetPage(buf);
        entry->nblocks++;
    }

    if (!cryo_decompress(COMP_LZ4, compressed, compressed_size, entry->data))
        return CRYO_ERR_DECOMPRESSION_FAILED;

    return CRYO_ERR_SUCCESS;
}

CryoError
cryo_read_data(Relation rel, BlockNumber blockno, CacheEntry *result)
{
    bool    found;
    PageId  pageId = {
        .relid = RelationGetRelid(rel),
        .blockno = blockno
    };
    PageIdCacheEntry *item;

     /* TODO: do not rely on RelationGetNumberOfBlocks; refer to metapage */
    if (RelationGetNumberOfBlocks(rel) <= blockno)
    {
        *result = InvalidCacheEntry;
        return CRYO_ERR_WRONG_STARTING_BLOCK;
    }

    /* check with hashtable */
    item = hash_search(pagemap, &pageId, HASH_FIND, &found);

    if (!found || item->entry == InvalidCacheEntry)
    {
        int i;
        CacheEntry  new_entry = InvalidCacheEntry;
        CryoError   err;
        TS          min_ts = -1; /* max unsigned long long */
        int         min_ts_pos;

        /* find available spot in cache or evict old cache entry */
        for (i = 0; i < CACHE_SIZE; ++i)
        {
            if (cache[i].ts == 0)
            {
                new_entry = i;
                break;
            }

            if (cache[i].ts < min_ts)
                min_ts_pos = i;
        }

        if (new_entry == InvalidCacheEntry)
        {
            /*
             * We didn't manage to find uninitialized cache entries. But we
             * found least recently used one. We also need to remove the old
             * record from pagemap.
             */
            elog(DEBUG1,
                 "pg_cryogen: evicted cache entry for (%i, %i)",
                 cache[min_ts_pos].key.relid,
                 cache[min_ts_pos].key.blockno);
            hash_search(pagemap, &cache[min_ts_pos].key, HASH_REMOVE, NULL);
            new_entry = min_ts_pos;
        }

        /* cache entry is not found, load data from disk */
        item = hash_search(pagemap, &pageId, HASH_ENTER, NULL);
        item->key.relid = RelationGetRelid(rel);
        item->key.blockno = blockno;
        item->entry = new_entry;

        /* load cryo block */
        cache[item->entry].ts = get_current_timestamp_ms();
        cache[item->entry].key = item->key;
        err = cryo_read_decompress(rel, blockno, &cache[new_entry]);
        if (err != CRYO_ERR_SUCCESS)
        {
            *result = item->entry = InvalidCacheEntry;
            return err;
        }
    }

    *result = item->entry;
    return CRYO_ERR_SUCCESS;
}

uint32
cryo_cache_get_pg_nblocks(CacheEntry entry)
{
    return cache[entry].nblocks;
}

char *
cryo_cache_get_data(CacheEntry entry)
{
    CacheEntryHeader *header = &cache[entry];

    /* TODO: check that entry is valid */

    /* update usage timestamp for LRU */
    header->ts = get_current_timestamp_ms();

    return header->data;
}

TransactionId
cryo_cache_get_xid(CacheEntry entry)
{
    return cache[entry].xid;
}

char *
cryo_cache_err(CryoError err)
{
    switch (err)
    {
        case CRYO_ERR_SUCCESS:
            return "success";
        case CRYO_ERR_WRONG_STARTING_BLOCK:
            return "wrong starting block number";
        case CRYO_ERR_DECOMPRESSION_FAILED:
            return "decompression failed";
        default:
            return "unknown error";
    }
}

