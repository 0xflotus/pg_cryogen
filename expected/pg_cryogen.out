CREATE EXTENSION pg_cryogen;
CREATE TEMPORARY TABLE tmp_data
AS SELECT gs as id, md5(gs::TEXT) as msg FROM generate_series(1, 500) as gs;
COPY tmp_data TO '/tmp/pg_cryogen_data.csv' WITH csv;
CREATE TABLE data (id INT4 NOT NULL, msg TEXT) USING pg_cryogen;
COPY data FROM '/tmp/pg_cryogen_data.csv' WITH csv;
SELECT * FROM data LIMIT 10;
 id |               msg                
----+----------------------------------
  1 | c4ca4238a0b923820dcc509a6f75849b
  2 | c81e728d9d4c2f636f067f89cc14862c
  3 | eccbc87e4b5ce2fe28308fd9f2a7baf3
  4 | a87ff679a2f3e71d9181a67b7542122c
  5 | e4da3b7fbbce2345d7772b0674a318d5
  6 | 1679091c5a880faf6fb5e6087eb1b2dc
  7 | 8f14e45fceea167a5a36dedd4bea2543
  8 | c9f0f895fb98ab9159f51fd0297e236d
  9 | 45c48cce2e2d7fbdea1afc51c7c6ad26
 10 | d3d9446802a44259755d38e6d163e820
(10 rows)

SELECT avg(id) FROM data;
         avg          
----------------------
 250.5000000000000000
(1 row)

SELECT count(*) FROM data;
 count 
-------
   500
(1 row)

CREATE INDEX data_id_idx ON data USING btree (id);
SET enable_seqscan = f;
ANALYZE data;
EXPLAIN (COSTS OFF) SELECT * FROM data WHERE id = 500;
              QUERY PLAN              
--------------------------------------
 Index Scan using data_id_idx on data
   Index Cond: (id = 500)
(2 rows)

SELECT * FROM data WHERE id = 500;
 id  |               msg                
-----+----------------------------------
 500 | cee631121c2ec9232f3a2f028ad5c89b
(1 row)

DROP INDEX data_id_idx;
CREATE INDEX data_id_idx ON data USING brin (id);
EXPLAIN (COSTS OFF) SELECT * FROM data WHERE id = 500;
               QUERY PLAN               
----------------------------------------
 Bitmap Heap Scan on data
   Recheck Cond: (id = 500)
   ->  Bitmap Index Scan on data_id_idx
         Index Cond: (id = 500)
(4 rows)

SELECT * FROM data WHERE id = 500;
 id  |               msg                
-----+----------------------------------
 500 | cee631121c2ec9232f3a2f028ad5c89b
(1 row)

SET enable_seqscan = t;
SET pg_cryogen.compression_method = 'lz4';
COPY data FROM '/tmp/pg_cryogen_data.csv' WITH csv;
SELECT count(*) FROM data;
 count 
-------
  1000
(1 row)

VACUUM data;
-- inserts
TRUNCATE data;
INSERT INTO data
SELECT gs as id, md5(gs::TEXT) as msg FROM generate_series(1, 500) as gs;
-- check FKs
DROP INDEX data_id_idx;
CREATE UNIQUE INDEX data_id_idx ON data USING btree (id);
CREATE TABLE ref_table (id SERIAL, data_id INT4 REFERENCES data (id));
INSERT INTO ref_table VALUES (1, 50), (2, 100), (3, 250); -- cryo_tuple_lock is called here
EXPLAIN (COSTS OFF) SELECT * FROM data JOIN ref_table ON data_id = data.id;
                 QUERY PLAN                 
--------------------------------------------
 Hash Join
   Hash Cond: (ref_table.data_id = data.id)
   ->  Seq Scan on ref_table
   ->  Hash
         ->  Seq Scan on data
(5 rows)

SELECT * FROM data JOIN ref_table ON data_id = data.id LIMIT 3;
 id  |               msg                | id | data_id 
-----+----------------------------------+----+---------
  50 | c0c7c76d30bd3dcaefc96f40275bdc0a |  1 |      50
 100 | f899139df5e1059396431415e770c6dd |  2 |     100
 250 | 6c9882bbac1c7093bd25041881277658 |  3 |     250
(3 rows)

DROP TABLE ref_table;
DROP INDEX data_id_idx;
-- check rescan
CREATE TABLE other_table AS
SELECT gs as id, '2000-01-01'::DATE + gs as dt FROM generate_series(1, 500) as gs;
SET enable_hashjoin = f;
SET enable_mergejoin = f;
SET enable_material = f;
EXPLAIN (COSTS OFF) SELECT * FROM data JOIN other_table USING (id);
                QUERY PLAN                 
-------------------------------------------
 Nested Loop
   Join Filter: (data.id = other_table.id)
   ->  Seq Scan on other_table
   ->  Seq Scan on data
(4 rows)

SELECT * FROM data JOIN other_table USING (id) LIMIT 3;
 id |               msg                |     dt     
----+----------------------------------+------------
  1 | c4ca4238a0b923820dcc509a6f75849b | 01-02-2000
(1 row)

